alphagov/paas-admin
# Enhance performance page

What
----

Present data in tables 

In addition to graphs we're showing data in tables for a11y.

While graph data is plotted directly for source in weekly increments, table data is shown as max number of <metric> per month for previous months, excluding current month as we already say "X currently <metric>"- and also latest value might not be the max number for the month, so there would be a mismatch.


Add functions to manipulate data

Data comes in weekly increments and that would have been a very long table for the entire year, so we've decided to show max numbers for each month.

To do that we have the `exportMaxPerMonthDataValues` function, that takes the source data, sorts it by descending value column,  and then removes duplicate monthly values, ensuring highest monthly value
remains. 
We could have used Lodash but native javascript works well.

For the organisation data we needed a different data structure as we're showing the data in two columns per row.

`combineMetrics` function takes in two arrays, utilises the `exportMaxPerMonthDataValues` to extract max monthly values
and finally combines data into a new structure.


To work with dates we have a `formatDate` function, as MomentJs is in [maintenance mode only(https://momentjs.com/docs/#/-project-status/) and we need to move away from its usage slowly.

We utilise native javascript to take in datetime stamp and output formatted string, in our case "monthName year"

How to review
-------------

- run `npm install` to update dependencies
- run `npm run stub-api` in first cli window
- run `npx serve` in second cli window to serve the sample metrics data 
- open third cli window and:
  - `export PLATFORM_METRICS_ENDPOINT=http://localhost:5000` (which should be what` npx serve` wlll run on - if not, amend)
  - run ` npm run start:with-stub-api` and visit the `/performance` page
- review code (best by comits)

Who can review
---------------

not @kr8n3r 

## Visual change

![visual page screenshot](https://user-images.githubusercontent.com/3758555/109832326-25419580-7c38-11eb-8fc4-6fa7e9cbcb0c.png)
